kernel/kernel-api/src/main/java/io/delta/kernel/internal/ScanImpl.java

```diff
@@ -309,4 +346,79 @@ private CloseableIterator<FilteredColumnarBatch> applyDataSkipping(
               filteredScanFileBatch.getData(), Optional.of(newSelectionVector));
         });
   }
+
+  /**
+   * Wraps a scan file iterator such that we emit {@link ScanReport} to the engine upon success and
+   * failure. Since most of our scan building code is lazily executed (since it occurs as
+   * maps/filters over an iterator) potential errors don't occur just within `getScanFile`s
+   * execution, but rather may occur as the returned iterator is consumed. Similarly, we cannot
+   * report a successful scan until the iterator has been fully consumed and the log read/filtered
+   * etc. This means we cannot report the successful scan within `getScanFiles` but rather must
+   * report after the iterator has been consumed.
+   *
+   * <p>This method takes a scan file iterator. It wraps the methods `next` and `hasNext` such that
+   * in the case of an exception originating from there, we will first report a failed ScanReport
+   * before propagating the exception. It also reports a success ScanReport when the iterator is
+   * closed, if and only if, all the elements have been consumed.
+   */
+  private CloseableIterator<FilteredColumnarBatch> wrapWithMetricsReporting(
+      CloseableIterator<FilteredColumnarBatch> scanIter, ScanReportReporter reporter) {
+    return new CloseableIterator<FilteredColumnarBatch>() {
+
+      @Override
+      public void close() throws IOException {
+        try {
+          if (!scanIter.hasNext()) {
+            // The entire scan file iterator has been successfully consumed, report successful Scan
+            reporter.reportSuccess();
+          } else {
+            // TODO what should we do if the iterator is not fully consumed?
```

Comment:
Will we ever be invoked/informed if the iterator is not fully consumed? I don't think there's anything we _can_ even do, right?

Reply:
I think answered below, but if the iterator is closed but not fully consumed (i.e. `hasNext = true`)

Reply:
great point, I missed this! Thanks!